Recompile WebGLBackground's material when the background changes

jsantell opened the pull request:
A follow up for the great work recently landed in r98 for cubemap support in backgrounds: https://github.com/mrdoob/three.js/pull/15187 https://github.com/mrdoob/three.js/pull/15186 https://github.com/mrdoob/three.js/pull/15081

Fixes #15325

This patch is necessary to accommodate new encodings, formats, etc. when a new texture is set on `scene.background`. `webgl_materials_envmaps_hdr.html` was updated to display the cubemap which can be used to reproduce this issue. Whatever is the first texture set determines the compiled encoding functions, so this example, if loading LDR first, results in HDR/RGBM16 images looking like:

![screenshot from 2018-11-21 13-34-28](https://user-images.githubusercontent.com/641267/48872093-83664980-ed9c-11e8-9380-a18eb6bd42c4.png)
![screenshot from 2018-11-21 13-34-52](https://user-images.githubusercontent.com/641267/48872099-85c8a380-ed9c-11e8-83db-e0448ebf9761.png)

The following patch fixes this issue and properly updates between HDR/LDR formats. I'd also suggest the change to the example since there aren't any other examples of HDR cubemaps being set on scene.background AFAICT.

jsantell committed:
Recompile WebGLBackground's material when the background changes to accomodate new encodings, formats, etc. Fixes #15325

WestLangley commented on src/renderers/webgl/WebGLBackground.js:
```diff
@@ -18,6 +18,7 @@ function WebGLBackground( renderer, state, objects, premultipliedAlpha ) {
 
 	var planeMesh;
 	var boxMesh;
+	var previousBackground;
```
We would probably call that `currentBackground`. It only needs to be reset when it changes -- not every frame.

jsantell replied:
:+1:

WestLangley commented on src/renderers/webgl/WebGLBackground.js:
```diff
@@ -85,6 +86,12 @@ function WebGLBackground( renderer, state, objects, premultipliedAlpha ) {
 
+			if ( background !== previousBackground ) {
+
+				boxMesh.material.needsUpdate = true;
```
What happens if the new texture has not loaded yet? Where is the new `texture.encoding` value assigned in the app? If it is in the texture loader callback, then that is a problem...

WestLangley commented:
Thanks for this. Code injection is a bit problematic. We may have to discuss this a bit...

WestLangley commented on examples/webgl_materials_envmaps_hdr.html:
```diff
@@ -114,7 +116,9 @@
-				new THREE.HDRCubeTextureLoader().load( THREE.UnsignedByteType, hdrUrls, function ( hdrCubeMap ) {
+				new THREE.HDRCubeTextureLoader().load( THREE.UnsignedByteType, hdrUrls, function ( _hdrCubeMap ) {
+
+					hdrCubeMap = _hdrCubeMap;
```
Can you please explain the reason for this pattern? Can you restore the former pattern and still get the demo to work?

jsantell replied:
Which pattern specifically? We do need to make the CubeTexture global to set to scene.background later on, and this just involved the least amount of changes -- happy to change it to anything else

jsantell commented:
@WestLangley thanks for the review! I've made some updates:

* Rename `previousBackground` to `currentTexture`, only apply if `!==`
* Clear out `currentTexture` if background is unset or a color.
* Also store `currentTextureVersion`, which we use to indicate if a change has occurred in the texture that would require recompiling the background material
* The example no longer waits for textures to be loaded and sets the scene.background immediately from the texture returned from the loader. The encoding gets set in the callback. This all seems to work as expected when throttling the network

How I tested this with the included example was with network throttling, commenting out the envmap generation in texture callbacks (they clobber scene.background when rendering cubemaps, which gave a false positive previously) and in the HDR texture callback, do `params.envMap = 'RGBM16'` -- this results in the following flow:

* The HDR background is set upfront -- renders all black, until the texture is loaded which will then display the HDR background.
* If added the `params.envMap = 'RGBM16'` in the HDR texture callback, or manually change it quickly, then the background will then display the RGBM16 texture (most likely not yet loaded) with LinearEncoding -- incorrect but expected considering the demo, as the encoding is set in the callback.
* Once the encoding is set on the texture and the RGBM texture has loaded, the incorrect LinearEncoding texture is swapped out with the correct encoding a moment later.
* Model will be all black due to disabling the PMREM/Envmaps to temporarily eliminate false positives

What do you think? Anything I'm missing?

jsantell commented:
There are no examples using a WebGLRenderTarget as a scene.background, I could make an example testing all the examples of scene.background (and reflection tests with #15302), not sure how WebGLRenderTarget would work here.

jsantell commented:
This does require compiling the shader multiple times a frame if using more than one scene with the same renderer. `WeakMap<Scene, { version, target }>` could solve this. WDYT?

jsantell commented:
Updated with better version matching by checking if it's a WebGLRenderTargetCube, and updated the EXR example to display as a background as well (using EquirectangularToCubeGenerator to test out WebGLRenderTargetCube). Render targets were working without this update, but now checks versioning if needed -- If using a render target, the material will only be recompiled if setting a `needsUpdate` on the texture, which seems appropriate.

**Caching**

This is looking good to me -- let me know if there's any interest in caching the materials mentioned https://github.com/mrdoob/three.js/pull/15299#issuecomment-441866944 -- in our use case (https://github.com/GoogleWebComponents/model-viewer) we use the same renderer for multiple scenes (not counting auxilliary scenes used in EquirectangularToCubeGenerator and PMREMGenerator)  which will mean recompiling the material multiple times a frame. I'm not sure what kind of overhead that is, but could be avoided with a WeakMap. Another possible solution could be making available all variants of `mapTexelToLinear` for the background texture. Happy to handle this in this PR, in a follow up, or not at all.

**Outstanding Questions**

* `scene.background` does not support `WebGLRenderTarget` correct?
* Any interest in a `scene.background` example that showcases all different variants?

WestLangley commented:
Thanks for working on this.

>This does require compiling the shader multiple times a frame if using more than one scene with the same renderer.

You do not have to solve all edge cases. Users can create their own skybox if need be.

>scene.background does not support WebGLRenderTarget correct?

I believe `scene.background` currently supports `WebGLRenderTarget.texture`.

>Any interest in a scene.background example that showcases all different variants?

Yes. Maybe you can update the existing example in a later PR.

Sorry, I do not have the time presently to focus on the remainder of your questions... Try to keep this PR simple. You can always file another PR later.

jsantell commented:
@WestLangley sounds great, thanks! Let me know if there's anything else needed for this unit of work to land and I can break up future work into smaller changes.

mrdoob commented on src/renderers/webgl/WebGLBackground.js:
```diff
@@ -82,9 +88,21 @@ function WebGLBackground( renderer, state, objects, premultipliedAlpha ) {
 			boxMesh.material.uniforms.tFlip.value = ( background.isWebGLRenderTargetCube ) ? 1 : - 1;
 
+			if ( currentBackground !== background ||
+				 ( currentBackground === background &&
```
I don't think this check is needed.

mrdoob commented on src/renderers/webgl/WebGLBackground.js:
```diff
@@ -133,6 +151,18 @@ function WebGLBackground( renderer, state, objects, premultipliedAlpha ) {
 			planeMesh.material.uniforms.uvTransform.value.copy( background.matrix );
 
+			if ( currentBackground !== background ||
+				 ( currentBackground === background &&
```
Same here.

mrdoob commented on src/renderers/webgl/WebGLBackground.js:
```diff
@@ -26,11 +30,13 @@ function WebGLBackground( renderer, state, objects, premultipliedAlpha ) {
 		if ( background === null ) {
 
 			setClear( clearColor, clearAlpha );
+			currentBackground = currentBackgroundVersion = null;
```
```js
currentBackground = null;
currentBackgroundVersion = null;
```

mrdoob commented on src/renderers/webgl/WebGLBackground.js:
```diff
@@ -18,6 +18,10 @@ function WebGLBackground( renderer, state, objects, premultipliedAlpha ) {
 	var boxMesh;
+	// Store the current background texture and its `version`
+	// so we can recompile the material accordingly.
+	var currentBackground;
```
```js
var currentBackground = null;
var currentBackgroundVersion = null;
```

mrdoob replied:
Actually, I think this should be more correct:

```js
var currentBackgroundVersion = 0;
```

mrdoob replied:
`texture.version = 0` means that it didn't load yet.

mrdoob commented on src/renderers/webgl/WebGLBackground.js:
```diff
@@ -26,11 +30,13 @@ function WebGLBackground( renderer, state, objects, premultipliedAlpha ) {
 
 			setClear( background, 1 );
 			forceClear = true;
+			currentBackground = currentBackgroundVersion = null;
```
```js
currentBackground = null;
currentBackgroundVersion = 0;
```

mrdoob commented:
Thanks!
